### DISCUSSION TITLE
Adding files to index through git_index_add_from_buffer

### DISCUSSION THREAD

**Discussion by Jamaslab** [libgit2/libgit2#6424] [Q&A]
Hello,
I am trying to use `git_index_add_from_buffer` to commit blob changes to a branch.
The following code works without issues:

```
git_index_entry* entry = git_index_get_bypath(index, path_to_file, 0);
git_index_add_from_buffer(index, entry, buffer, buffer_length);
```

However, this works only if the blob object is already part of the index. 

According to documentation ([git_index_add_from_buffer](https://libgit2.org/libgit2/#HEAD/group/index/git_index_add_from_buffer)):
```
This method will create a blob in the repository that owns the index and then add the index entry to the index. The path of the entry represents the position of the blob relative to the repository's root folder.

If a previous index entry exists that has the same path as the given 'entry', it will be replaced. Otherwise, the 'entry' will be added.
```

So I wrote the following:
```
git_index_entry entry;
entry.mode = GIT_FILEMODE_BLOB;
entry.path = path_to_file;
git_index_add_from_buffer(index, &entry, buffer, buffer_length);
```

However, when I run this code, `git_index_add_from_buffer `returns -1. What am I doing wrong? Am I supposed to create the git_index_entry in a different way?

Thanks!

**Comment by kcsaul** [comment]
Hi @Jamaslab,

Does the file exist in the working directory? (if there is one)

If not, you probably need to set the entry flags before adding.  In my project, I add files to the index without a working directory, for which I found I needed to set `entry.flags = GIT_INDEX_ADD_FORCE | GIT_INDEX_ADD_CHECK_PATHSPEC;`
